### PR TITLE
Clarify the distinction of unset scalar fields vs unset message fields.

### PR DESCRIPTION
**Pull request by smolkaj**
There is a subtle distinction between the treatment of default-valued message fields vs default-valued scalar fields. This has bitten me and others at Google one to many times, so this PR clarifies the distinction. 

This PR is not intended to introduce any semantic changes.

### CODE REVIEW

File: docs/v1/P4Runtime-Spec.mdk [hunk 1]
```diff
@@ -5394,13 +5408,22 @@ it populates the `MasterArbitrationUpdate` message using the `device_id`,
 also populates the `status` field in the `MasterArbitrationUpdate` according to
 the rules in an [earlier section](#sec-mastership-updates).
 
-The sender need not specify an `election_id`. If the `election_id` is not
-specified, the sender's `election_id` is considered lower than any
+### Unset Election ID
+
+The sender need not specify an `election_id`. If the `election_id` is
+unset, the sender's `election_id` is considered lower than any
```
Comment (chrispsommers) [inline docs/v1/P4Runtime-Spec.mdk:5414]: Just out of curiosity, if two controllers have unset election ids, does that mean they are both standbys, and nether can take control?

**Review by chrispsommers — APPROVED** [review]
LGTM

File: docs/v1/P4Runtime-Spec.mdk [hunk 2]
```diff
@@ -2044,13 +2044,35 @@ When writing the config via a `SetForwardingPipelineConfig` RPC, the cookie
...
+In Protobuf version 3 (*proto3*), the default value of scalar fields is `0` for
+numeric types such as `int32`, and the empty string `""` for string types
+(`string` and `bytes`). An application, such as the P4Runtime client or
+server, is **unable to distinguish** between an unset scalar field and a scalar
+field set to its default value. Therefore, we usually reserve the default
+values 0 and "" of scalar fields to mean "unset".
+
+In particular, 0 is not a valid P4 object ID and it
+is an error to specify 0 for any P4 object ID in a non-read request towards the
+switch, such as in a `WriteRequest` or a `SetForwardingPipelineConfigRequest`.
```
Comment (antoninbas) [inline docs/v1/P4Runtime-Spec.mdk:2065]: nit: I know you just moved that paragraph, but "server" would be a more generic & appropriate term than "switch"

Comment (smolkaj) [reply docs/v1/P4Runtime-Spec.mdk]: Good point. Done!

File: docs/v1/P4Runtime-Spec.mdk (hunk 1 above)
Comment (smolkaj) [reply docs/v1/P4Runtime-Spec.mdk:5414]: Correct, you can have any number of standbys. A standby can become primary by sending another election request with a higher election ID.

**Comment by antoninbas** [comment]
let's hold off on merging this, I think @jafingerhut wanted to take a look over the weekend

**Comment by smolkaj** [comment]
Sounds great, no rush. Thanks everyone for the reviews!

File: docs/v1/P4Runtime-Spec.mdk [hunk 3]
```diff
@@ -5637,12 +5659,21 @@ it populates the `MasterArbitrationUpdate` message using the `device_id`,
...
-specified, the sender's `election_id` is considered lower than any
-`election_id`, and the sender will thus never become primary. This way, a
-controller can choose to be a backup controller, in order to avoid
-"flapping" (if a backup controller connects to the switch shortly before the
-actual primary controller, therefore becoming primary temporarily).
+### Unset Election ID
+
+The sender need not specify an `election_id`. If the `election_id` is
+unset, the sender's `election_id` is considered lower than any
+`election_id`, and the sender will thus never become master. This way, a
```
Comment (jafingerhut) [inline docs/v1/P4Runtime-Spec.mdk:5666]: Wasn't there a pass through the spec a while back to replace occurrences of `master` with `primary`?  This change seems to undo some of that.

**Review by jafingerhut — APPROVED** [review]
LGTM, with possible exception of primary->master replacements (see comment).

File: docs/v1/P4Runtime-Spec.mdk (hunk 3 above)
Comment (antoninbas) [reply docs/v1/P4Runtime-Spec.mdk:5666]: thanks for catching this Andy. @smolkaj could you take care of this, I am not sure why you removed instances of `primary` in your PR?

Comment (antoninbas) [reply docs/v1/P4Runtime-Spec.mdk:5666]: @smolkaj friendly ping

Comment (smolkaj) [reply docs/v1/P4Runtime-Spec.mdk:5666]: Ugh... looks like maybe my change was based on an old version of the spec? 

Let me try to revert & reapply the changes.

Comment (smolkaj) [reply docs/v1/P4Runtime-Spec.mdk]: It was easy to fix actually.

**Comment by jafingerhut** [comment]
As of commit 6 on this PR, I see all "master" occurrences replaced with "primary", so looks ready to go to me.

File: docs/v1/P4Runtime-Spec.mdk [hunk 4]
```diff
@@ -2082,13 +2082,35 @@ When writing the config via a `SetForwardingPipelineConfig` RPC, the cookie
...
+In Protobuf version 3 (*proto3*), the default value of scalar fields is `0` for
+numeric types such as `int32`, and the empty string `""` for string types
+(`string` and `bytes`). An application, such as the P4Runtime client or
+server, is **unable to distinguish** between an unset scalar field and a scalar
+field set to its default value. Therefore, we usually reserve the default
+values 0 and "" of scalar fields to mean "unset".
+
+In particular, 0 is not a valid P4 object ID and it
+is an error to specify 0 for any P4 object ID in a non-read request towards the
+server, such as in a `WriteRequest` or a `SetForwardingPipelineConfigRequest`.
```
[the remark this answers is not in the document]
Comment (smolkaj) [reply docs/v1/P4Runtime-Spec.mdk:2103]: Removed, I didn't keep it intentionally. If I removed and then reverted, that was unintentional.

**Review by antoninbas — APPROVED** [review]

**Comment by smolkaj** [comment]
Thanks!

**Pull request merged**